README.md
<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/p0bs/shiny.moduler/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/p0bs/shiny.moduler/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

<br/>


`shiny.moduler` provides a simple Shiny app that controls underlying Shiny modules. As I'm seeking to show the Shiny functionality, the rest of the package is deliberately unpolished, with hardly any unit tests or error checks.

Why create this package? Well, I found that most of the code examples for Shiny stopped just before the point of building a main module to control sub-modules. As such, I thought I would go a bit further, in the hope that it might prove useful to others.

To see the resulting Shiny app, [run it here](https://robin.shinyapps.io/useshiny/).

<br/>

## Installation

You can install the latest version of `shiny.moduler` as follows:

``` r
devtools::install_github("p0bs/shiny.moduler")
```

<br/>

## Example

To run the app that controls and contains the sub-modules, use:

``` r
library(shiny.moduler)
app_main()
```

To run the sub-modules on a standalone basis, use:

``` r
library(shiny.moduler)
app_sectors(standalone = TRUE)
app_indicators(standalone = TRUE)
```


NEWS.md
# shiny.moduler 0.2.3
* Tweaking code to remove NOTES from R-CMD-CHECK

# shiny.moduler 0.2.2
* Changing app address

# shiny.moduler 0.2.1
* Adding some missing double colons


# shiny.moduler 0.2.0
* Adding some polish to UI and outputs


# shiny.moduler 0.1.0
* Added a `NEWS.md` file to track changes to the package.
* Initial simple functionality


tests/testthat.R
library(testthat)
library(shiny.moduler)

test_check("shiny.moduler")


tests/testthat/test-get_data_sectors.R
data_test_sectors <- sectors_data

test_that("Expected column names", {
  expect_equal(
    colnames(data_test_sectors), 
    c("Date", "Sector", "Return"))
})


tests/testthat/test-get_data_indicators.R
data_test_indicators <- shiny.moduler:::get_data_indicators()

test_that("Expected column names", {
  expect_equal(
    colnames(data_test_indicators), 
    c("date", "series", "level")
    )
})


R/main.R
#' @title Running the main app that controls the shiny sub modules
#' @description This function runs the main app that controls the shiny sub-modules.
#' @keywords module
#' @examples
#' \dontrun{
#' app_main()
#' }
app_main <- function(){
  
  data_import <- shiny.moduler::sectors_data
  data_economy <- get_data_indicators()
  
  ui <- bslib::page_fillable(
    bslib::card(
      full_screen = TRUE,
      bslib::card_header("Demonstrating the use of modules"),
      bslib::layout_sidebar(
        sidebar = bslib::sidebar(
          shiny::selectInput("analysis", "Analysis", choices = c("Sector Market History", "Economic Indicators"), multiple = FALSE, selected = "Economic Indicators"),
          shiny::uiOutput("inputs", fill = "item")),
        shiny::uiOutput("outputs"))),
    title = "Testing modules", 
    theme = bslib::bs_theme(
      bootswatch = "minty",
      version = 5, bg = "white", fg = "black",
      primary = "#63431c", secondary = "#f9f3ec",
      code_font = bslib::font_collection("Arial", "sans-serif"),
      "dropdown-bg" = "#f9f3ec",
      "dropdown-link-hover-bg" = "#63431c",
      "card-border-color" = "white",
      "navbar-bg" = "#f9f3ec"
      )
    )
  
  server <- function(input, output) {
    
    ui_inputs <- shiny::reactive({
      switch (
        input$analysis,
        "Sector Market History" = ui_sectors(id = "sectors", data_import = data_import),
        "Economic Indicators" = ui_indicators(id = "indicators"))
    })
    
    output$inputs <- shiny::renderUI(ui_inputs())
    
    ui_outputs <- shiny::reactive({
      switch (
        input$analysis,
        "Sector Market History" = server_sectors(id = "sectors", data_import = data_import),
        "Economic Indicators" = server_indicators(id = "indicators", data_import = data_economy))
    })
    
    output$outputs <- shiny::renderUI(ui_outputs())
  }
  
  shiny::shinyApp(ui, server)
  
}


R/sectors.R
#' @name ui_sectors
#' @title Components relating to the sector performance module
#' @description This function provides the components for running the sector performance module, be that as a sub-module of a bigger app or on a standalone basis.
#' @param id The identifier used in shiny to namespace a module, thereby linking the ui and server components
#' @param data_import The data imported on sector performance
#' @param standalone Whether to view the module on a standalone basis (TRUE) or not (FALSE, the default)
#' @keywords module
#' @examples
#' \dontrun{
#' ui_sectors(id = "sectors", data_import = data_imported, standalone = TRUE)
#' }
NULL

#' @name server_sectors
#' @title Components relating to the sector performance module
#' @description This function provides the components for running the sector performance module, be that as a sub-module of a bigger app or on a standalone basis.
#' @param id The identifier used in shiny to namespace a module, thereby linking the ui and server components
#' @param data_import The data imported on sector performance
#' @keywords module
#' @examples
#' \dontrun{
#' server_sectors(id = "sectors", data_import = data_imported)
#' }
#' 
#' @importFrom rlang .data
#' 
NULL

#' @name app_sectors
#' @title Components relating to the sector performance module
#' @description This function provides the components for running the sector performance module, be that as a sub-module of a bigger app or on a standalone basis.
#' @keywords module
#' @examples
#' \dontrun{
#' app_sectors()
#' }
NULL

#' @rdname ui_sectors
ui_sectors <- function(id, data_import, standalone = FALSE){
  if(standalone){
    shiny::tagList(
      shiny::sidebarPanel(
        shiny::selectInput(shiny::NS(id, "year_start"), "Starting Year", choices = unique(lubridate::year(data_import$Date)), selected = min(lubridate::year(data_import$Date), na.rm = TRUE)),
        shiny::selectInput(shiny::NS(id, "year_end"), "Final Year", choices = unique(lubridate::year(data_import$Date)), selected = max(lubridate::year(data_import$Date), na.rm = TRUE)),
        shiny::radioButtons(shiny::NS(id, "button_output"), "", choices = c("Plot", "Table"), selected = "Table")
      ),
      shiny::mainPanel(
        shiny::uiOutput(shiny::NS(id, "output_sectors"))
      )
    )
  } else {
    shiny::tagList(
      shiny::selectInput(shiny::NS(id, "year_start"), "Starting Year", choices = unique(lubridate::year(data_import$Date)), selected = min(lubridate::year(data_import$Date), na.rm = TRUE)),
      shiny::selectInput(shiny::NS(id, "year_end"), "Final Year", choices = unique(lubridate::year(data_import$Date)), selected = max(lubridate::year(data_import$Date), na.rm = TRUE)),
      shiny::radioButtons(shiny::NS(id, "button_output"), "", choices = c("Plot", "Table"), selected = "Table")
    )
  }
}

#' @rdname server_sectors
server_sectors <- function(id, data_import){
  shiny::moduleServer(id, function(input, output, session){
    
    stopifnot(!shiny::is.reactive(data_import))
    
    summary_sectors <- shiny::reactive({
      data_import %>% 
        dplyr::filter(
          .data$Date >= lubridate::make_date(year = input$year_start),
          .data$Date <= lubridate::make_date(year = input$year_end, month = 12L, day = 31L)
        ) %>% 
        dplyr::group_by(.data$Sector) %>% 
        dplyr::summarise(
          mean = mean(.data$Return, na.rm = TRUE), 
          stdev = stats::sd(.data$Return, na.rm = TRUE)
        )
    })
    
    output_rendered <- shiny::reactive({
      switch (input$button_output,
              "Table" = shiny::renderTable({
                summary_sectors() %>%
                  dplyr::mutate(
                    mean = .data$mean * 100,
                    stdev = .data$stdev * 100
                  ) %>% 
                  dplyr::arrange(dplyr::desc(.data$mean), dplyr::desc(.data$stdev))
              }),
              "Plot" = shiny::renderPlot({
                summary_sectors() %>%
                  ggplot2::ggplot(
                    ggplot2::aes(
                      x = .data$stdev,
                      y = .data$mean,
                      colour = .data$Sector
                    )
                  ) + 
                  ggplot2::geom_point() + 
                  ggplot2::scale_x_continuous(labels = scales::percent_format(accuracy = 0.01)) +
                  ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 0.01)) +
                  ggplot2::theme_minimal() + 
                  ggplot2::theme(plot.title.position = "plot") +
                  ggplot2::labs(
                    title = "Comparing the typical return and risk of sectors",
                    subtitle = "\nTypical daily return",
                    x = "\n Variability of daily returns",
                    y = NULL, 
                    colour = NULL
                  )
              })
      )
    })
    
    output$output_sectors <- shiny::renderUI(
      output_rendered()
      )
    
  })
}

#' @rdname app_sectors
app_sectors <- function(){
  
  data_import <- shiny.moduler::sectors_data
  
  ui <- shiny::fluidPage(
    shiny::titlePanel("Testing Modules with Open Financial Data"),
    ui_sectors(id = "sectors", data_import = data_import, standalone = TRUE)
    )
  
  server <- function(input, output, session) {
    server_sectors(id = "sectors", data_import = data_import)
  }
  
  shiny::shinyApp(ui, server)
  
}


R/indicators.R
#' @name ui_indicators
#' @title Components relating to the economic indicators module
#' @description This function provides the components for running the economic indicators module, be that as a sub-module of a bigger app or on a standalone basis.
#' @param id The identifier used in shiny to namespace a module, thereby linking the ui and server components
#' @param standalone Whether to view the module on a standalone basis (TRUE) or not (FALSE, the default)
#' @keywords module
#' @examples
#' \dontrun{
#' ui_indicators(id = "indicators", standalone = TRUE)
#' }
#' 
NULL

#' @name server_indicators
#' @title Components relating to the economic indicators module
#' @description This function provides the components for running the economic indicators module, be that as a sub-module of a bigger app or on a standalone basis.
#' @param id The identifier used in shiny to namespace a module, thereby linking the ui and server components
#' @param data_import The data imported on economic indicators
#' @keywords module
#' @examples
#' \dontrun{
#' server_indicators(id = "indicators", data_import = data_imported)
#' }
#'
#' @importFrom rlang .data
#' 
NULL

#' @name app_indicators
#' @title Components relating to the economic indicators module
#' @description This function provides the components for running the economic indicators module, be that as a sub-module of a bigger app or on a standalone basis.
#' @keywords module
#' @examples
#' \dontrun{
#' app_indicators()
#' }
#' 
NULL

#' @rdname ui_indicators
ui_indicators <- function(id, standalone = FALSE){
  if(standalone){
    shiny::tagList(
      shiny::sidebarPanel(
        shiny::radioButtons(shiny::NS(id, "button_output"), "", choices = c("Plot", "Table"), selected = "Table")
      ),
      shiny::mainPanel(
        shiny::uiOutput(shiny::NS(id, "output_indicators"))
      )
    )
  } else {
    shiny::tagList(
      shiny::radioButtons(shiny::NS(id, "button_output"), "", choices = c("Plot", "Table"), selected = "Table")
    )
  }
}

#' @rdname server_indicators
server_indicators <- function(id, data_import){
  shiny::moduleServer(id, function(input, output, session){
    
    stopifnot(!shiny::is.reactive(data_import))
    
    output_rendered <- shiny::reactive({
      switch (input$button_output,
              "Table" = shiny::renderTable({
                data_import %>%
                  tibble::as_tibble() %>% 
                  dplyr::group_by(.data$series) %>% 
                  dplyr::slice_max(order_by = .data$date, n = 1) %>% 
                  dplyr::ungroup() %>% 
                  dplyr::mutate(
                    date = as.character.Date(.data$date),
                    latest = .data$level * 100
                    ) %>% 
                  dplyr::select(-.data$level)
              }),
              "Plot" = shiny::renderPlot({
                data_import %>% 
                  feasts::autoplot(.vars = .data$level) +
                  ggplot2::theme_minimal() + 
                  ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
                  ggplot2::theme(plot.title.position = "plot") +
                  ggplot2::labs(
                    title = "Comparing changes in economic activity and implied long-term prices \n",
                    x = NULL,
                    y = NULL
                  )
              })
      )
    })
    
    output$output_indicators <- shiny::renderUI(
      output_rendered()
      )
    
  })
}

#' @rdname app_indicators
app_indicators <- function(){
  
  data_economy <- get_data_indicators()
  
  ui <- shiny::fluidPage(
    shiny::titlePanel("Testing Modules with Open Economic Indicators"),
    ui_indicators(id = "indicators", standalone = TRUE)
    )
  
  server <- function(input, output, session) {
    server_indicators(id = "indicators", data_import = data_economy)
  }
  
  shiny::shinyApp(ui, server)
  
}


R/get_data_sectors.R
#' @title Imports data relating to the performance of US sectors over time
#' @description This function enables you to import data relating to the performance of US sectors over time.
#' @keywords data
#' @examples
#' \dontrun{
#' get_data_sectors()
#' }
get_data_sectors <- function(){
  link_17 <- 'http://mba.tuck.dartmouth.edu/pages/faculty/ken.french/ftp/17_Industry_Portfolios_daily_CSV.zip'
  temp <- tempfile()
  utils::download.file(link_17 ,temp)
  data_import <- readr::read_csv(
    temp, 
    col_types = readr::cols(`...1` = readr::col_date(format = "%Y%m%d")), 
    skip = 9, 
    na = c(-99.99, -999, "NA", "N/A", "n/a", "na")
    ) %>% 
    dplyr::group_by(`...1`) %>% 
    dplyr::slice(1) %>% 
    dplyr::ungroup() %>% 
    dplyr::mutate(
      dplyr::across(.cols = tidyselect::where(is.numeric), .fns = ~ .x/100)
    ) %>% 
    dplyr::rename("Date" = `...1`) %>% 
    tidyr::gather(key = "Sector", value = "Return", -Date)
  
  return(data_import)
  
}


R/get_data_indicators.R
#' @title Imports data relating to key US economic indicators over time
#' @description This function enables you to import data relating to key US economic indicators over time.
#' @keywords data
#' @examples
#' \dontrun{
#' get_data_indicators()
#' }
#'
#' @importFrom rlang .data
get_data_indicators <- function(){
  
  data_economy <- tidyquant::tq_get(
    x = c(
      "T10YIE",  # 10 Yr Breakeven
      "OB000334Q" # Real Y-O-Y GDP
    ),
    get = "economic.data",
    from = "2002-01-01"
  ) |> 
    dplyr::mutate(
      series = dplyr::if_else(
        .data$symbol == "T10YIE", 
        "10 Yr Breakeven", 
        "Real Y-O-Y GDP"
      ),
      level = .data$price / 100
    ) |> 
    dplyr::select(-.data$symbol, -.data$price) |> 
    tsibble::as_tsibble(key = .data$series, index = .data$date) 

  return(data_economy)
  
}


R/app.R
library(feasts)
library(lubridate)
library(shiny)
library(tidyquant)
library(tidyverse)
library(tsibble)

data_economy <- tq_get(
  x = c(
    "T10YIE",  # 10 Yr Breakeven
    "OB000334Q" # Real Y-O-Y GDP
  ),
  get = "economic.data",
  from = "1990-01-01"
) %>% 
  mutate(
    series = if_else(
      symbol == "T10YIE", 
      "10 Yr Breakeven", 
      "Real Y-O-Y GDP"
    ),
    level = price / 100
  ) %>% 
  select(-symbol, -price) %>% 
  as_tsibble(key = series, index = date) %>%
  filter(date > '2002-01-01')

# Sector data
link_17 <- 'http://mba.tuck.dartmouth.edu/pages/faculty/ken.french/ftp/17_Industry_Portfolios_daily_CSV.zip'
temp <- tempfile()
download.file(link_17 ,temp)
data_import <- read_csv(temp, skip = 9, na = c(-99.99, -999, "NA", "N/A", "n/a", "na")) %>% 
  mutate(
    across(.cols = is.numeric, .fns = ~ .x/100),
    Date = ymd(as.integer(`...1`)), 
    .before = 1
  ) %>% 
  select(-`...1`) %>% 
  gather(key = "Sector", value = "Return", -Date)

ui <- fluidPage(
    titlePanel("Testing Modules"),
    fluidPage(
      selectInput("analysis", "Analysis", choices = c("Sectors", "Indicators"), multiple = FALSE, selected = "Indicators"),
      uiOutput("inputs"),
      uiOutput("outputs"))
    )

server <- function(input, output) {
  
  ui_inputs <- reactive({
    switch (
      input$analysis,
      "Sectors" = ui_sectors(id = "sectors", data_import = data_import),
      "Indicators" = ui_indicators(id = "indicators"))
  })
  
  output$inputs <- renderUI(ui_inputs())
  
  ui_outputs <- reactive({
    switch (
      input$analysis,
      "Sectors" = server_sectors(id = "sectors", data_import = data_import),
      "Indicators" = server_indicators(id = "indicators", data_import = data_economy))
  })
  
  output$outputs <- renderUI(ui_outputs())
}

# Run the application 
shinyApp(ui = ui, server = server)
